### DISCUSSION TITLE
Do I have to define data before I use them?

### DISCUSSION THREAD

**Discussion by sophief10** [jspsych/jsPsych#2175] [Q&A] [answered]
Hi everyone, 

I am trying to create an experiment in which participants use a Likert scale to respond to an image. Here is the code:
```

 var pre_task = {
            timeline: [
                {
                    type: 'survey-likert',
                    preamble: jsPsych.timelineVariable('image'),
                    questions: [
                        {
                            labels: likert_scale, 
                            prompt: "blah blah"
                        }
                    ],
                    on_finish: function(data) {
                        return data.response.Q0;
                    },
                    data: {
                        foodImage: jsPsych.timelineVariable('food'),
                        rating: data.response.Q0,
                        segment: "pre-task"   
                    }
                }
            ],
            timeline_variables: [
            { image:'<img src="image address"></img>'},
            { image:'<img src="image address"></img>'},
            { image:'<img src="image address"></img>'}
            ],
            randomize_order: true
        };
```

The problem is that, when I try to run this code, my debug console tells me that "data" is undefined and, therefore, the code fails. This issue confuses me because I thought that "data" was a predetermined parameter in jsPsych, i.e. that the computer will assume that "data" exists even if the coder does not explicitly define it. That said, even if the data parameter is not predetermined, I think I still defined it as an object in my code. Therefore, I do not understand why my code will not work.

Has anyone run into an issue like this before? If so, do you know how to resolve it?

**Comment by jodeleeuw** [comment, marked as answer]
Hi @woodenchair,

You are getting an error because of this line:

```js
	rating: data.response.Q0
```

You are right that `data` is a universal plugin parameter in jsPsych, but in this case you are trying to reference the value of `data.response` before the trial has happened. At the time that your definition of this timeline is created there is no data yet, and so the JavaScript interpreter doesn't know what the variable `data` refers to here.

It looks like what you are trying to do is basically rename `data.response.Q0` to `rating` in the data for the trial. You can do this in the `on_finish` parameter, like this:

```js
on_finish: function(data){
	data.rating = data.response.Q0;
}
```

The reason that this works in `on_finish` is that `on_finish` is a function and this function doesn't execute until the trial finishes. The function's `data` argument gets the object of data for the trial, and you can both modify it (e.g., set the value of `data.rating`) and access values from within it (e.g., get the value of `data.response.Q0`).

**Reply by sophief10** [reply]
Makes sense! Thanks for your help, @jodeleeuw !
